File: gitdot-web/packages/api/src/endpoint/repository/get-repository-blobs.ts
import { z } from "zod";
import { RepositoryBlobsResource } from "../../resource";
import type { Endpoint } from "../endpoint";

export const GetRepositoryBlobsRequest = z.object({
  ref_name: z.string().optional(),
  paths: z.array(z.string()),
});
export type GetRepositoryBlobsRequest = z.infer<
  typeof GetRepositoryBlobsRequest
>;

export const GetRepositoryBlobsResponse = RepositoryBlobsResource;
export type GetRepositoryBlobsResponse = z.infer<
  typeof GetRepositoryBlobsResponse
>;

export const GetRepositoryBlobs = {
  path: "/repository/{owner}/{repo}/blobs",
  method: "POST",
  request: GetRepositoryBlobsRequest,
  response: GetRepositoryBlobsResponse,
} as const satisfies Endpoint;
export type GetRepositoryBlobs = typeof GetRepositoryBlobs;

File: gitdot-api/src/endpoint/repository/get_repository_blobs.rs
use serde::{Deserialize, Serialize};

use super::default_ref;
use crate::{endpoint::Endpoint, resource::repository::RepositoryBlobsResource};

fn default_refs() -> Vec<String> {
    vec![default_ref()]
}

pub struct GetRepositoryBlobs;

impl Endpoint for GetRepositoryBlobs {
    const PATH: &'static str = "/repository/{owner}/{repo}/blobs";
    const METHOD: http::Method = http::Method::POST;

    type Request = GetRepositoryBlobsRequest;
    type Response = GetRepositoryBlobsResponse;
}

#[derive(ApiRequest, Debug, Serialize, Deserialize)]
pub struct GetRepositoryBlobsRequest {
    #[serde(default = "default_refs")]
    pub refs: Vec<String>,
    pub paths: Vec<String>,
}

pub type GetRepositoryBlobsResponse = RepositoryBlobsResource;
